scripts/build_shapes.py

```diff
@@ -705,6 +734,340 @@ def _process_func_download_pop(c_code):
+            window_topleft_longitude,
+            affine_transform[3],
+            affine_transform[4],
+            window_topleft_latitude,
+        )
+
+    # Set an empty numpy array with the dimensions of the country .tif file
+    # np_map_ID will contain a ID for each location (undefined is 0)
+    # ID corresponds to a specific geometry in country_rows
+    np_map_ID = np.zeros((y_axis_len, x_axis_len))
+
+    # List to contain the mappings of id to GADM_ID
+    id_to_GADM_ID = []
+
+    # Loop the country_rows geoDataFrame
+    for i in range(len(country_rows)):
+        # Set the current geometry
+        cur_geometry = country_rows.iloc[i]["geometry"]
+
+        # In windowed mode we check if bounds of geometry overlap the window
+        if windowed:
+            latitude_min = cur_geometry.bounds[1]
+            latitude_max = cur_geometry.bounds[3]
+
+            # In the following cases we don't have to continue the loop
+            # If the geometry is above the window
+            if latitude_min > window_topleft_latitude:
+                continue
+            # If the geometry is below the window
+            if latitude_max < window_botright_latitude:
+                continue
+
+        # Generate a mask for the specific geometry
+        temp_mask = rasterio.features.geometry_mask(
+            [cur_geometry],
+            (y_axis_len, x_axis_len),
+            affine_transform,
+            invert=True,
+            all_touched=True,
+        )
+
+        # Map the values of counter value to np_map_ID
+        np_map_ID[temp_mask] = i + 1
+
+        # Store the id -> GADM_ID mapping
+        id_to_GADM_ID.append([i + 1, country_rows.iloc[i]["GADM_ID"]])
+
+    # Return np_map_ID as type 'H' np.ushort
+    # 'H' -> https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.ushort
+    # This lowers memory usage, note: ID has to be within the range [0,65535]
+    return np_map_ID.astype("H"), pd.DataFrame(id_to_GADM_ID).set_index(0)
+
+
+def compute_population(country_rows, WorldPop_inputfile, out_logging=False):
+    """
+    Function computes the population for the given country rows
+
+    Inputs:
+        -------
+        country_rows:
+
+        WorldPop_inputfile:
+
+
+    Outputs:
+        --------
+        df_pop_count: Dataframe with columns
+            - "pop" containing population of GADM_ID region
+            - "GADM_ID"
+
+
+    """
+    # Get the features of the worldpop input file
+    transform, worldpop_dim = get_worldpop_features(WorldPop_inputfile)
+
+    worldpop_y_dim, worldpop_x_dim = worldpop_dim
+
+    # Rasterio doesn't support lower than float32 readout
+    # Hence reading the file will take up: nbytes = 4 * y_dim * x_dim
+    expected_bytes_input_read = 4 * worldpop_y_dim * worldpop_x_dim
+
+    # Introduce a max byte size to avoid overfilling RAM
+    # Ensure worldpop_byte_limit > 883 * 10**6 (minimum memory for 'US')
+    worldpop_byte_limit = 3096 * 10**6
```

Comment:
I agree it makes sense to take this from the snakemake file, just one consideration is how to divide the allocated memory effectively once we introduce more cores.

There is a minimum amount of memory required for loading the largest window. The largest window in our dataset is the 'US' as it spans the entire globe and as such has a minimum of 883 MB.
In addition, the `worldpop_byte_limit` targets the most memory intensive operation (src.read(1, ...)) but the whole code has overhead for storing the other arrays, hence I don't want to dilute `mem_mb` if its intended as a concrete ceiling.

It's possible to test and find the size of the overhead, but conservatively I estimated 1GB, so theoretically 4GB per process with `worldpop_byte_limit = 3096 * 10**6`.

A simple approach would be:

`worldpop_byte_limit = max(883, (mem_mb / nProcess) - 1000)`

Obviously this would cause problems when nProcess is large as it would likely result in memory overflow (assuming the computer doesn't have 2GB * nProcess of RAM available). What do you think @davide-f ?

Reply:
Thanks for the comment!
So, I agree that the memory in the snakefile shall be the top required by all tasks.
That number can be changed anyway. 800mb can be increased.
Moreover for simplicity, I would span them uniformly across different process. No need for the -1000 overhead

Reply:
@GridGrapher I've seen you are finalizing the PR :) Do you have any comment here?

Reply:
Implemented in 418207e28d023382afe3c21460e08c9595e2d822